Failed to CREATE/RENAME when create blocked by ownership and renaming

If you create a DatabaseUser object with an existing username, that'll fail and sit (like it should). However, if you update DatabaseUser's username to something that doesn't conflict, you get an error because it tries to CREATE and looks up the wrong stuff:

renaming username from manage-ui (conflict) to manage-ui-secret:

2022-07-25T23:26:28.535Z INFO controller.databaseuser Starting workers {"reconciler group": "mysql.apps.cuppett.dev", "reconciler kind": "DatabaseUser", "worker count": 1}
2022-07-25T23:26:29.015Z INFO controllers.DatabaseUser Taking ownership of the authentication secret. {"Name": "manage-secret-pw", "Namespace": "mariadb-test"}
2022/07/25 23:28:20 /workspace/orm/model.go:84 record not found
[19.084ms] [rows:0] SELECT * FROM `mysql`.`user` WHERE user = 'manage-ui-secret' ORDER BY `user`.`Host` LIMIT 1
2022/07/25 23:28:20 /workspace/orm/model.go:84 record not found
[0.006ms] [rows:0]
2022-07-25T23:28:20.156Z INFO controllers.DatabaseUser User does not exist or failed retrieving {"Host": "hq.cuppett.com", "Name": "manage-ui"}
2022/07/25 23:28:20 /workspace/controllers/databaseuser_controller.go:532 Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'BY '8tbCr0C3pO2tJnO@n#IY1eQp'' at line 1
[4.727ms] [rows:0] CREATE USER 'manage-ui' IDENTIFIED WITH 'mysql_native_password' BY '8tbCr0C3pO2tJnO@n#IY1eQp'
2022-07-25T23:28:20.161Z ERROR controllers.DatabaseUser Failed to execute query. {"Host": "hq.cuppett.com", "Name": "manage-ui", "Query": "CREATE USER 'manage-ui' IDENTIFIED WITH 'mysql_native_password' BY '8tbCr0C3pO2tJnO@n#IY1eQp'", "error": "Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'BY '8tbCr0C3pO2tJnO@n#IY1eQp'' at line 1"}
github.com/cuppett/mysql-dba-operator/controllers.(*DatabaseUserReconciler).userCreate
/workspace/controllers/databaseuser_controller.go:270
github.com/cuppett/mysql-dba-operator/controllers.(*DatabaseUserReconciler).Reconcile
/workspace/controllers/databaseuser_controller.go:185
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227
2022/07/25 23:28:20 /workspace/orm/model.go:84 record not found
[19.173ms] [rows:0] SELECT * FROM `mysql`.`user` WHERE user = 'manage-ui-secret' ORDER BY `user`.`Host` LIMIT 1
2022/07/25 23:28:20 /workspace/orm/model.go:84 record not found
[0.005ms] [rows:0]
2022-07-25T23:28:20.191Z INFO controllers.DatabaseUser User does not exist or failed retrieving {"Host": "hq.cuppett.com", "Name": "manage-ui"}
2022/07/25 23:28:20 /workspace/controllers/databaseuser_controller.go:532 Error 1396: Operation CREATE USER failed for 'manage-ui'@'%'
[5.723ms] [rows:0] CREATE USER 'manage-ui'
2022-07-25T23:28:20.197Z ERROR controllers.DatabaseUser Failed to execute query. {"Host": "hq.cuppett.com", "Name": "manage-ui", "Query": "CREATE USER 'manage-ui'", "error": "Error 1396: Operation CREATE USER failed for 'manage-ui'@'%'"}
github.com/cuppett/mysql-dba-operator/controllers.(*DatabaseUserReconciler).userCreate
/workspace/controllers/databaseuser_controller.go:270
github.com/cuppett/mysql-dba-operator/controllers.(*DatabaseUserReconciler).Reconcile
/workspace/controllers/databaseuser_controller.go:185
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227

--- a/api/v1alpha1/adminconnection_types.go
+++ b/api/v1alpha1/adminconnection_types.go
@@ -344,18 +344,25 @@ func (in *AdminConnection) DatabaseMine(gormDB *gorm.DB, database *Database) boo
 func (in *AdminConnection) UserMine(gormDB *gorm.DB, user *DatabaseUser) bool {
 
 	var managedUser orm.ManagedUser
-
-	// If it doesn't exist, go ahead and take it!
-	if orm.UserExists(gormDB, user.Spec.Username) == nil {
-		return true
+	names := make([]string, 0, 2)
+	names = append(names, user.Spec.Username)
+	if user.Spec.Username != user.Status.Username && user.Status.Username != "" {
+		names = append(names, user.Status.Username)
 	}
 
-	// If it does exist, let's check the triple after fetching by UID
-	gormDB.Limit(1).Find(&managedUser, "uuid = ?", string(user.UID))
-	if managedUser.Username == user.Spec.Username &&
-		managedUser.Name == user.Name &&
-		managedUser.Namespace == user.Namespace {
-		return true
+	for _, name := range names {
+		// If it doesn't exist, go ahead and take it!
+		if orm.UserExists(gormDB, name) == nil {
+			continue
+		}
+
+		// If it does exist, let's check the triple after fetching by UID
+		gormDB.Limit(1).Find(&managedUser, "uuid = ?", string(user.UID))
+		if managedUser.Username != name ||
+			managedUser.Name != user.Name ||
+			managedUser.Namespace != user.Namespace {
+			return false
+		}
 	}
-	return false
+	return true
 }

--- a/api/v1alpha1/databaseuser_types.go
+++ b/api/v1alpha1/databaseuser_types.go
@@ -24,7 +24,6 @@ import (
 // DatabaseUserSpec defines the desired state of DatabaseUser
 type DatabaseUserSpec struct {
 	AdminConnection AdminConnectionRef `json:"adminConnection"`
-	// TODO: Block or allow the rename of a user (currently would just CREATE new one if changed)
 	// +kubebuilder:validation:MaxLength:=32
 	// +kubebuilder:validation:MinLength:=1
 	Username string `json:"username"`

--- a/config/crd/bases/mysql.apps.cuppett.dev_databaseusers.yaml
+++ b/config/crd/bases/mysql.apps.cuppett.dev_databaseusers.yaml
@@ -108,8 +108,6 @@ spec:
                     type: boolean
                 type: object
               username:
-                description: 'TODO: Block or allow the rename of a user (currently
-                  would just CREATE new one if changed)'
                 maxLength: 32
                 minLength: 1
                 type: string

--- a/controllers/databaseuser_controller.go
+++ b/controllers/databaseuser_controller.go
@@ -161,7 +161,7 @@ func (r *DatabaseUserReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, err
 	}
 
-	if loop.instance.Status.Username == "" {
+	if loop.instance.Status.Username == "" && loop.adminConnection.UserMine(loop.db, loop.instance) {
 		// Ensuring old/new username is always set.
 		loop.instance.Status.Username = loop.instance.Spec.Username
 		err = r.Status().Update(ctx, loop.instance)
@@ -185,7 +185,7 @@ func (r *DatabaseUserReconciler) Reconcile(ctx context.Context, req ctrl.Request
 			r.Log.Error(err, "Failure adding the finalizer.", "Name",
 				loop.instance.Name, "Namespace", loop.instance.Namespace)
 		}
-	} else {
+	} else if loop.instance.Status.Username != "" {
 		exists, err := r.userExists(&loop)
 
 		if !exists {
@@ -198,6 +198,8 @@ func (r *DatabaseUserReconciler) Reconcile(ctx context.Context, req ctrl.Request
 			if updated {
 				loop.instance.Status.SyncTime = metav1.NewTime(time.Now())
 			}
+		} else {
+			loop.instance.Status.Message = "Ownership failed. Unable to update user."
 		}
 
 		if err != nil {
@@ -205,6 +207,9 @@ func (r *DatabaseUserReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		} else {
 			err = r.Status().Update(ctx, loop.instance)
 		}
+	} else if loop.instance.Status.Username == "" {
+		loop.instance.Status.Message = "Invalid username specified."
+		err = r.Status().Update(ctx, loop.instance)
 	}
 	return ctrl.Result{}, err
 }
@@ -245,7 +250,7 @@ func (r *DatabaseUserReconciler) secretOwnershipOk(loop *UserLoopContext) bool {
 
 func (r *DatabaseUserReconciler) userExists(loop *UserLoopContext) (bool, error) {
 
-	user := orm.UserExists(loop.db, loop.instance.Spec.Username)
+	user := orm.UserExists(loop.db, loop.instance.Status.Username)
 
 	if user == nil {
 		r.Log.Info("User does not exist or failed retrieving", "Host", loop.adminConnection.Spec.Host,
@@ -314,6 +319,21 @@ func (r *DatabaseUserReconciler) userUpdate(ctx context.Context, loop *UserLoopC
 			"Old", loop.instance.Status.Username, "New", loop.instance.Spec.Username)
 		loop.instance.Status.Username = loop.instance.Spec.Username
 		loop.instance.Status.Message = "User renamed"
+
+		managedUser := orm.ManagedUser{
+			Uuid:      string(loop.instance.UID),
+			Namespace: loop.instance.Namespace,
+			Name:      loop.instance.Name,
+			Username:  loop.instance.Spec.Username,
+		}
+
+		tx := loop.db.Updates(&managedUser)
+		if tx.Error != nil {
+			r.Log.Error(tx.Error, "Failed to update managed user record.", "Host", loop.adminConnection.Spec.Host, "Name",
+				loop.instance.Spec.Username)
+		}
+		tx.Commit()
+
 		return true, nil
 	}
 

--- a/controllers/databaseuser_controller_test.go
+++ b/controllers/databaseuser_controller_test.go
@@ -94,13 +94,15 @@ var _ = Describe("DatabaseUser", func() {
 				return userObject.Status.Message
 			}).WithContext(ctx).Should(Equal("Created user"))
 
-			// Fetch a fresh DatabaseUser
-			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
-			Expect(err).ToNot(HaveOccurred())
-
-			databaseUser.Spec.Username = "test-user-renamed"
-			err = k8sClient.Update(ctx, databaseUser)
-			Expect(err).ToNot(HaveOccurred())
+			// Continue to do this until we stop getting 409.
+			Eventually(func() error {
+				// Fetch a fresh DatabaseUser
+				err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
+				Expect(err).ToNot(HaveOccurred())
+				databaseUser.Spec.Username = "test-user-renamed"
+				err = k8sClient.Update(ctx, databaseUser)
+				return err
+			}).WithContext(ctx).Should(BeNil())
 
 			Eventually(func() string {
 				userObject := &DatabaseUser{}
@@ -149,14 +151,15 @@ var _ = Describe("DatabaseUser", func() {
 				return userObject.Status.Message
 			}).WithContext(ctx).Should(Equal("Created user"))
 
-			// Fetch a fresh DatabaseUser
-			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
-			Expect(err).ToNot(HaveOccurred())
-
-			// Update the username field
-			databaseUser.Spec.Username = "existing_user"
-			err = k8sClient.Update(ctx, databaseUser)
-			Expect(err).ToNot(HaveOccurred())
+			// Continue to do this until we stop getting 409.
+			Eventually(func() error {
+				// Fetch a fresh DatabaseUser
+				err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
+				Expect(err).ToNot(HaveOccurred())
+				databaseUser.Spec.Username = "existing_user"
+				err = k8sClient.Update(ctx, databaseUser)
+				return err
+			}).WithContext(ctx).Should(BeNil())
 
 			Eventually(func() string {
 				userObject := &DatabaseUser{}
@@ -206,14 +209,15 @@ var _ = Describe("DatabaseUser", func() {
 				return userObject.Status.Message
 			}).WithContext(ctx).Should(Equal("Invalid username specified."))
 
-			// Fetch a fresh DatabaseUser
-			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
-			Expect(err).ToNot(HaveOccurred())
-
-			// Update the username field
-			databaseUser.Spec.Username = "switch-good-user"
-			err = k8sClient.Update(ctx, databaseUser)
-			Expect(err).ToNot(HaveOccurred())
+			// Continue to do this until we stop getting 409.
+			Eventually(func() error {
+				// Fetch a fresh DatabaseUser
+				err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
+				Expect(err).ToNot(HaveOccurred())
+				databaseUser.Spec.Username = "switch-good-user"
+				err = k8sClient.Update(ctx, databaseUser)
+				return err
+			}).WithContext(ctx).Should(BeNil())
 
 			Eventually(func() string {
 				userObject := &DatabaseUser{}

--- a/controllers/databaseuser_controller_test.go
+++ b/controllers/databaseuser_controller_test.go
@@ -95,7 +95,6 @@ var _ = Describe("DatabaseUser", func() {
 			}).WithContext(ctx).Should(Equal("Created user"))
 
 			// Fetch a fresh DatabaseUser
-			databaseUser = &DatabaseUser{}
 			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
 			Expect(err).ToNot(HaveOccurred())
 
@@ -151,7 +150,6 @@ var _ = Describe("DatabaseUser", func() {
 			}).WithContext(ctx).Should(Equal("Created user"))
 
 			// Fetch a fresh DatabaseUser
-			databaseUser = &DatabaseUser{}
 			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
 			Expect(err).ToNot(HaveOccurred())
 
@@ -209,7 +207,6 @@ var _ = Describe("DatabaseUser", func() {
 			}).WithContext(ctx).Should(Equal("Invalid username specified."))
 
 			// Fetch a fresh DatabaseUser
-			databaseUser = &DatabaseUser{}
 			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
 			Expect(err).ToNot(HaveOccurred())
 
@@ -226,7 +223,6 @@ var _ = Describe("DatabaseUser", func() {
 			}).WithContext(ctx).Should(Equal("Created user"))
 
 			// Fetching again and checking the actual username.
-			databaseUser = &DatabaseUser{}
 			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
 			Expect(err).ToNot(HaveOccurred())
 			Expect(databaseUser.Status.Username).To(Equal("switch-good-user"))

--- a/controllers/databaseuser_controller_test.go
+++ b/controllers/databaseuser_controller_test.go
@@ -1,6 +1,7 @@
 package controllers
 
 import (
+	"github.com/cuppett/mysql-dba-operator/orm"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	. "github.com/onsi/ginkgo/v2"
@@ -15,11 +16,9 @@ var _ = Describe("DatabaseUser", func() {
 
 	Describe("Creation Scenario", func() {
 
-		var databaseUser *DatabaseUser
-
 		It("Happy Path", func(ctx SpecContext) {
 
-			databaseUser = &DatabaseUser{
+			databaseUser := &DatabaseUser{
 				ObjectMeta: metav1.ObjectMeta{
 					Name:      "test-user",
 					Namespace: ServerAdminConnection.Namespace,
@@ -48,4 +47,190 @@ var _ = Describe("DatabaseUser", func() {
 		}, NodeTimeout(time.Second*30))
 	})
 
+	Describe("Rename Scenario", func() {
+
+		It("Should have an existing user to rename", func(ctx SpecContext) {
+			// Pre-reqs
+			cache := make(map[types.UID]*orm.ConnectionDefinition)
+			gormDB, err := ServerAdminConnection.GetDatabaseConnection(ctx, k8sClient, cache)
+			Expect(err).ToNot(HaveOccurred())
+
+			// Create database user to be renamed
+			createQuery := "CREATE USER '" + Escape("existing_user") + "'"
+			tx := gormDB.Exec(createQuery)
+			Expect(tx.Error).To(BeNil())
+		})
+
+		It("Happy Path", func(ctx SpecContext) {
+
+			databaseNamespacedName := types.NamespacedName{
+				Name:      "test-user-name",
+				Namespace: ServerAdminConnection.Namespace,
+			}
+
+			databaseUser := &DatabaseUser{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-user-name",
+					Namespace: ServerAdminConnection.Namespace,
+				},
+				Spec: DatabaseUserSpec{
+					AdminConnection: AdminConnectionRef{
+						Name: ServerAdminConnection.Name,
+					},
+					Username: "test-user-name",
+				},
+			}
+			err := k8sClient.Create(ctx, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(func() string {
+				userObject := &DatabaseUser{}
+				databaseNamespacedName := types.NamespacedName{
+					Namespace: databaseUser.Namespace,
+					Name:      databaseUser.Name,
+				}
+				err := k8sClient.Get(ctx, databaseNamespacedName, userObject)
+				Expect(err).ToNot(HaveOccurred())
+				return userObject.Status.Message
+			}).WithContext(ctx).Should(Equal("Created user"))
+
+			// Fetch a fresh DatabaseUser
+			databaseUser = &DatabaseUser{}
+			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+
+			databaseUser.Spec.Username = "test-user-renamed"
+			err = k8sClient.Update(ctx, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(func() string {
+				userObject := &DatabaseUser{}
+				err := k8sClient.Get(ctx, databaseNamespacedName, userObject)
+				Expect(err).ToNot(HaveOccurred())
+				return userObject.Status.Message
+			}).WithContext(ctx).Should(Equal("User renamed"))
+
+			// Fetching again and checking the actual username.
+			databaseUser = &DatabaseUser{}
+			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(databaseUser.Status.Username).To(Equal("test-user-renamed"))
+
+		}, NodeTimeout(time.Second*30))
+
+		It("Fail Rename To Existing User", func(ctx SpecContext) {
+			databaseNamespacedName := types.NamespacedName{
+				Name:      "orig-user-name",
+				Namespace: ServerAdminConnection.Namespace,
+			}
+
+			databaseUser := &DatabaseUser{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "orig-user-name",
+					Namespace: ServerAdminConnection.Namespace,
+				},
+				Spec: DatabaseUserSpec{
+					AdminConnection: AdminConnectionRef{
+						Name: ServerAdminConnection.Name,
+					},
+					Username: "orig-user-name",
+				},
+			}
+			err := k8sClient.Create(ctx, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(func() string {
+				userObject := &DatabaseUser{}
+				databaseNamespacedName := types.NamespacedName{
+					Namespace: databaseUser.Namespace,
+					Name:      databaseUser.Name,
+				}
+				err := k8sClient.Get(ctx, databaseNamespacedName, userObject)
+				Expect(err).ToNot(HaveOccurred())
+				return userObject.Status.Message
+			}).WithContext(ctx).Should(Equal("Created user"))
+
+			// Fetch a fresh DatabaseUser
+			databaseUser = &DatabaseUser{}
+			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+
+			// Update the username field
+			databaseUser.Spec.Username = "existing_user"
+			err = k8sClient.Update(ctx, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(func() string {
+				userObject := &DatabaseUser{}
+				err := k8sClient.Get(ctx, databaseNamespacedName, userObject)
+				Expect(err).NotTo(HaveOccurred())
+				return userObject.Status.Message
+			}).WithContext(ctx).Should(Equal("Ownership failed. Unable to update user."))
+
+			// Fetching again and checking the actual username.
+			databaseUser = &DatabaseUser{}
+			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(databaseUser.Status.Username).To(Equal("orig-user-name"))
+
+		}, NodeTimeout(time.Second*30))
+
+		It("Allow Rename To Non-Existent User", func(ctx SpecContext) {
+
+			databaseNamespacedName := types.NamespacedName{
+				Name:      "switch-good-user",
+				Namespace: ServerAdminConnection.Namespace,
+			}
+
+			databaseUser := &DatabaseUser{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "switch-good-user",
+					Namespace: ServerAdminConnection.Namespace,
+				},
+				Spec: DatabaseUserSpec{
+					AdminConnection: AdminConnectionRef{
+						Name: ServerAdminConnection.Name,
+					},
+					Username: "existing_user",
+				},
+			}
+			err := k8sClient.Create(ctx, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(func() string {
+				userObject := &DatabaseUser{}
+				databaseNamespacedName := types.NamespacedName{
+					Namespace: databaseUser.Namespace,
+					Name:      databaseUser.Name,
+				}
+				err := k8sClient.Get(ctx, databaseNamespacedName, userObject)
+				Expect(err).ToNot(HaveOccurred())
+				return userObject.Status.Message
+			}).WithContext(ctx).Should(Equal("Invalid username specified."))
+
+			// Fetch a fresh DatabaseUser
+			databaseUser = &DatabaseUser{}
+			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+
+			// Update the username field
+			databaseUser.Spec.Username = "switch-good-user"
+			err = k8sClient.Update(ctx, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(func() string {
+				userObject := &DatabaseUser{}
+				err := k8sClient.Get(ctx, databaseNamespacedName, userObject)
+				Expect(err).NotTo(HaveOccurred())
+				return userObject.Status.Message
+			}).WithContext(ctx).Should(Equal("Created user"))
+
+			// Fetching again and checking the actual username.
+			databaseUser = &DatabaseUser{}
+			err = k8sClient.Get(ctx, databaseNamespacedName, databaseUser)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(databaseUser.Status.Username).To(Equal("switch-good-user"))
+
+		}, NodeTimeout(time.Second*30))
+	})
 })

